### PR TITLE
sl-21: Authorization plugin

### PR DESCRIPTION
**Pull request by dimapopovych**
decided to look for a review and fix change requests while waiting for jwt PR fixes

### CODE REVIEW

File: backend/src/libs/packages/white-routes-config/libs/types/white-route.type.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { type HttpMethod } from '~/libs/packages/http/http.js';
+
+type WhiteRoute = {
+  routerPath: string;
+  httpMethods: HttpMethod[];
```
Comment (what1s1ove) [inline backend/src/libs/packages/white-routes-config/libs/types/white-route.type.ts]:
```suggestion
  methods: HttpMethod[];
```

File: backend/src/libs/packages/white-routes-config/white-routes-config.package.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  public getWhiteRoutes(): WhiteRoutes {
+    return this.whiteRoutes;
+  }
+
+  private createWhiteRoute(...paths: string[]): string {
+    return `/api/v1${paths.join('')}`;
+  }
+
+  private initWhiteRoutes(): WhiteRoutes {
+    return [
```
Comment (what1s1ove) [inline backend/src/libs/packages/white-routes-config/white-routes-config.package.ts]: just create a constant for it in app pacakge

File: backend/src/libs/plugins/authorization.plugin.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import fp from 'fastify-plugin';
+
+import { type HttpMethod } from '../packages/http/http.js';
+import { type WhiteRoutesConfig } from '../packages/white-routes-config/white-routes-config.js';
+
+const authorization = fp(
+  (
+    fastify: FastifyInstance,
+    { whiteRoutesConfig }: { whiteRoutesConfig: WhiteRoutesConfig },
+    done: () => void,
```
Comment (what1s1ove) [inline backend/src/libs/plugins/authorization.plugin.ts]:
```suggestion
```
prefer promise interface

File: backend/src/libs/plugins/authorization.plugin.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            (whiteRoute) =>
+              whiteRoute.routerPath === request.routerPath &&
+              whiteRoute.httpMethods.includes(request.method as HttpMethod),
+          );
+
+        if (isWhiteRoute) {
+          return;
+        }
+
+        await reply.code(403).send('TBA');
```
Comment (what1s1ove) [inline backend/src/libs/plugins/authorization.plugin.ts]:
```suggestion
        await reply.code(HttpCode.CODE);
```

File: backend/src/libs/packages/server-application/server-app.ts [hunk 5]
```diff
@@ -74,6 +86,12 @@ class ServerApp implements IServerApp {
     this.addRoutes(routers);
   }
 
+  private async initPlugins(): Promise<void> {
+    await this.app.register(authorization, {
+      whiteRoutesConfig: this.whiteRoutesConfig,
```
Comment (what1s1ove) [inline backend/src/libs/packages/server-application/server-app.ts]:
```suggestion
      whiteRoutes: // just pass a constant here
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/libs/packages/token/token.package.ts [hunk 6]
```diff
@@ -17,8 +18,17 @@ class Token implements IToken {
...
+    try {
+      const { payload } = await jwtVerify(token, this.createSecret());
+
+      return payload;
+    } catch {
+      throw new HttpError({
+        message: 'Invalid token',
+        status: HttpCode.UNAUTHORIZED,
+      });
+    }
```
Comment (what1s1ove) [inline backend/src/libs/packages/token/token.package.ts]:
```suggestion
```
do not an aditional try catch block

File: backend/src/libs/packages/server-application/libs/types/white-routes.type.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import { type WhiteRoute } from './white-route.type.js';
+
+type WhiteRoutes = WhiteRoute[];
```
Comment (what1s1ove) [inline backend/src/libs/packages/server-application/libs/types/white-routes.type.ts]:
```suggestion
```
do not an additional type. add `[]` where you need an array

File: backend/src/libs/plugins/authorization.plugin.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+import {
+  type HttpMethod,
+  HttpCode,
+  HttpError,
+} from '~/libs/packages/http/http.js';
+import { type WhiteRoutes } from '~/libs/packages/server-application/server-application.js';
+import { token } from '~/libs/packages/token/token.js';
+import { userService } from '~/packages/users/users.js';
+
+const authorization = fp(
```
Comment (what1s1ove) [inline backend/src/libs/plugins/authorization.plugin.ts]: provide user service as parameter for plugin

File: backend/src/libs/packages/token/libs/token.interface.ts [hunk 9]
```diff
@@ -1,6 +1,8 @@
+import { type JWTPayload } from 'jose';
+
 interface IToken {
   create(payload: unknown): Promise<string>;
-  verify(): void;
-  decode(): void;
+  verify(token: string): Promise<JWTPayload>;
```
Comment (what1s1ove) [inline backend/src/libs/packages/token/libs/token.interface.ts]:
```suggestion
  verify<T>(token: string): Promise<JWTPayload & T>;
```
should be generic. we need to provide all properties that we add to jwt payload

File: backend/src/libs/plugins/authorization.plugin.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+          whiteRoute.methods.includes(request.method as HttpMethod),
+      );
+
+      if (isWhiteRoute) {
+        return;
+      }
+
+      const [, requestToken] = request.headers.authorization?.split(' ') ?? [];
+
+      const { id } = await token.verify(requestToken);
```
Comment (what1s1ove) [inline backend/src/libs/plugins/authorization.plugin.ts]:
```suggestion
      const { userId } = await token.decode(requestToken);
```
change the jwn payload

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: backend/src/libs/plugins/authorization.plugin.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    fastify: FastifyInstance,
+    { whiteRoutesConfig }: { whiteRoutesConfig: WhiteRoutes },
+  ) => {
+    fastify.decorateRequest('user', null);
+
+    fastify.addHook('onRequest', async (request: FastifyRequest) => {
+      const isWhiteRoute = whiteRoutesConfig.some(
+        (whiteRoute) =>
+          whiteRoute.routerPath === request.routerPath &&
+          whiteRoute.methods.includes(request.method as HttpMethod),
```
Comment (andanf-e) [inline backend/src/libs/plugins/authorization.plugin.ts]: More readable
```suggestion
        (whiteRoute) => {
           const isWhiteRoute = whiteRoute.routerPath === request.routerPath;
           const isAllowedMethod = whiteRoute.methods.includes(request.method as HttpMethod);
           
           return isWhiteRoute && isAllowedMethod;
        }
```

File: backend/src/libs/packages/token/token.package.ts (hunk 6 above)
Comment (dimapopovych) [reply backend/src/libs/packages/token/token.package.ts]: Am I supposed to add handler for JWSInvalid exception into common error handler or just handle it inside auth plugin and throw HttpError?

Comment (what1s1ove) [reply backend/src/libs/packages/token/token.package.ts]: do not handle it for it. we will do it in the separate task

File: backend/src/libs/plugins/authorization.plugin.ts (hunk 8 above)
Comment (dimapopovych) [reply backend/src/libs/plugins/authorization.plugin.ts]: Should I create something like services obj and pass it to ServerApp constructor or I can just import it inside ServerApp?

Comment (what1s1ove) [reply backend/src/libs/plugins/authorization.plugin.ts]: just import in ServerApp

File: backend/src/libs/plugins/authorization/authorization.plugin.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        return isWhitePath && isAllowedMethod;
+      });
+
+      if (isWhiteRoute) {
+        return;
+      }
+
+      const [, requestToken] = request.headers.authorization?.split(' ') ?? [];
+
+      const { userId } = token.decode<{ userId: number }>(requestToken);
```
Comment (andanf-e) [inline backend/src/libs/plugins/authorization/authorization.plugin.ts:51]: TS does not complain on it, but it actually looks like you can have requestToken equal to `undefined`

Comment (dimapopovych) [reply backend/src/libs/plugins/authorization/authorization.plugin.ts:51]: Would be better to just specify default value or do something like
```
  const authHeader = request.headers.authorization;
  if (!authHeader) {
    throw;
  }
  const [, requestToken] = authHeader.split(' ');
  if (!requestToken) {
    throw
  }
```

Comment (andanf-e) [reply backend/src/libs/plugins/authorization/authorization.plugin.ts:51]: Yep, handling by throwing error is nice

**Review by andanf-e — APPROVED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**